share/nisar/examples/noise_spectra_analysis.py

```diff
@@ -0,0 +1,501 @@
+
+    Parameters
+    ----------
+    raw_data: complex np.ndarray
+        complex input raw data
+    cpi_len: int
+        Number of pulses to be used to estimate noise power of a single pulse
+
+    Returns
+    -------
+    raw_data_selcted: 2D complex array
+        Selected receive pulses for sample covariance estimation and Eigenvalue
+        Decomposition
+    """
+
+    num_pulses = raw_data.shape[0]
+
+    if cpi_len > num_pulses:
+        raise ValueError(f"Requested cpi_len={cpi_len} exceeds number of pulses {num_pulses}")
+    
+    # Equally spaced indices (without repetition)
+    indices = np.linspace(0, num_pulses - 1, cpi_len, dtype=int)
+    raw_data_selected = raw_data[indices, :]
+
+
+    return raw_data_selected
+
+
+def estimate_noise_pwr_pulse(
+    raw_data: np.ndarray, 
+    cpi_len: int = 6
+) -> float:
+    """Estimate noise power per pulse in the time domain using minimum Eigenvalue
+
+    Parameters
+    ----------
+    raw_data: complex np.ndarray
+        complex input raw data
+    cpi_len: int
+        Number of pulses to be used to estimate noise power of a single pulse
+        Default: 6 
+
+    Returns
+    -------
+    noise_pwr_pulse_est_db: float
+        Estmated noise power per pulse using minimum Eigenvalue
+    """
+
+    data_estimate = select_pulses_for_estimation(raw_data, cpi_len)  # [cpi_len, num_range]
+    sample_cov = data_estimate @ data_estimate.conj().T / data_estimate.shape[1]  # [cpi_len x cpi_len]
+    eig_vals, _ = la.eigh(sample_cov)  # ascending order for Hermitian matrix
+
+    noise_pwr_pulse_est_db = 10 * np.log10(eig_vals[0])
+
+    return noise_pwr_pulse_est_db
+
+def update_rfi_hit_stats(
+    rfi_mask_blk, 
+    rfi_freq_bin_hit_count, 
+    rfi_max_streak
+):
+    """
+    Update total RFI hit count and max streak for each frequency bin.
+
+    Parameters
+    ----------
+    rfi_mask_blk : np.ndarray (num_pulses_blk, num_fft)
+        Boolean mask indicating RFI detection for each pulse and frequency bin.
+    rfi_freq_bin_hit_count : np.ndarray (num_fft,)
+        Accumulator for total number of RFI hits per frequency bin.
+    rfi_max_streak : np.ndarray (num_fft,)
+        Accumulator for maximum consecutive hit streak per frequency bin.
+    """
+
+    # Update total hit count
+    rfi_freq_bin_hit_count += np.sum(rfi_mask_blk, axis=0)
+
+    # Vectorized max streak computation per frequency bin
+    diff = np.diff(np.pad(rfi_mask_blk.astype(int), ((1, 1), (0, 0)), mode='constant'), axis=0)
+
+    # starts and ends are tuples: starts[0]: pulse idx; starts[1]: freq bin idx
+    starts = np.where(diff == 1)
+    ends = np.where(diff == -1)
+
+    # For each frequency bin, compute max streak
+    for b in range(rfi_mask_blk.shape[1]):
+        start_idxs = starts[0][starts[1] == b]
+        end_idxs = ends[0][ends[1] == b]
+
+        if len(start_idxs) != len(end_idxs):
+            # If a streak starts but never ends (shouldn't happen with padding)
+             continue
```

Comment:
```suggestion
        assert len(start_idxs) == len(end_idxs), "expected an end to every beginning"
```
If it really won't happen, then I think an assertion is a good way to document it.

BTW the vectorization in this function is pretty impressive.